### PR TITLE
Add Pulse/Inngest router code example

### PR DESCRIPTION
**Pull request by sneub**

### CODE REVIEW

File: pulse/inngest-router/.env.sample [hunk 1]
```diff
@@ -0,0 +1,3 @@
+DATABASE_URL=""
```
Comment (jharrell) [inline pulse/inngest-router/.env.sample]: We generally call this `.env.example`

File: pulse/inngest-router/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+FROM node:20-alpine
```
Comment (jharrell) [inline pulse/inngest-router/Dockerfile:1]: We generally don't include dockerfiles in this repo. Is this needed?

File: pulse/inngest-router/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+### Fork or clone this repo
+
+You should operate with your own version of this service in a GitHub repository
+that you own. This allows you to track changes, and deploy to cloud platforms
+easily.
+
+### Set up your API keys
+
+Make a copy of the `.env.sample` file and update it with your own keys and
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
Make a copy of the `.env.example` file and update it with your own keys and
```

File: pulse/inngest-router/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+database connection string:
+
+```bash
+cp .env.sample .env
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
cp .env.example .env
```

File: pulse/inngest-router/README.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+your dashboard
+1. Choose to deploy from a **GitHub Repo**
+1. Authorize your GitHub account, and choose the GitHub repository you used in
+previous steps
+1. Add the respective environment variables:
+    1. DATABASE_URL
+    1. PULSE_API_KEY
+    1. INNGEST_EVENT_KEY
+
+### Run manually with Docker
```
Comment (jharrell) [inline pulse/inngest-router/README.md:82]: see above comments re: docker

File: pulse/inngest-router/package.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    "ts-node": "^10.9.2",
+    "typescript": "^5.5.4"
+  },
+  "dependencies": {
+    "@prisma/client": "^5.18.0",
+    "@prisma/extension-pulse": "^1.2.0",
+    "@types/node": "^22.3.0",
+    "dotenv": "^16.4.5",
+    "inngest": "^3.22.3"
+  },
```
Comment (jharrell) [inline pulse/inngest-router/package.json]:
```suggestion
  "devDependencies": {
    "@cloudflare/workers-types": "4.20240806.0",
    "nodemon": "3.1.4",
    "prisma": "5.18.0",
    "ts-node": "10.9.2",
    "typescript": "5.5.4"
  },
  "dependencies": {
    "@prisma/client": "5.18.0",
    "@prisma/extension-pulse": "1.2.0",
    "@types/node": "22.3.0",
    "dotenv": "16.4.5",
    "inngest": "3.22.3"
  },
```

File: pulse/inngest-router/prisma/schema.prisma [hunk 7]
```diff
@@ -0,0 +1,27 @@
+// This is your Prisma schema file,
+// learn more about it in the docs: https://pris.ly/d/prisma-schema
+
+// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
+// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init
+
+generator client {
+  provider      = "prisma-client-js"
+  binaryTargets = ["native", "linux-musl"]
```
Comment (jharrell) [inline pulse/inngest-router/prisma/schema.prisma:9]: what is `linux-musl` needed for?

File: pulse/inngest-router/prisma/schema.prisma [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+model Notification {
+  id      String @id @default(cuid())
```
Comment (jharrell) [inline pulse/inngest-router/prisma/schema.prisma]: I generally prefer autoincrement int IDs for our examples.

File: pulse/inngest-router/prisma/schema.prisma (hunk 7 above)
Comment (sneub) [reply pulse/inngest-router/prisma/schema.prisma:9]: For the docker build

File: pulse/inngest-router/Dockerfile (hunk 2 above)
Comment (sneub) [reply pulse/inngest-router/Dockerfile:1]: The goal of the example is to be a deployable bundle, rather than an artefact for learning -- so I'd prefer to keep it if it doesn't break too many rules?

File: pulse/inngest-router/README.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+- A **Pulse API key** which you can get by enabling Pulse in a project in your
+  [Prisma Data Platform](https://pris.ly/pdp) account (learn more in the
+  [docs](https://www.prisma.io/docs/platform/concepts/environments#api-keys))
+- An existing [Inngest](https://www.inngest.com) account and
+  [Event Key](https://www.inngest.com/docs/events/creating-an-event-key?ref=environment-variables)
+
+## Get started
+
+### Fork or clone this repo
```
Comment (jharrell) [inline pulse/inngest-router/README.md:16]: I think we should add instructions here on how to clone.

File: pulse/inngest-router/README.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+You should operate with your own version of this service in a GitHub repository
+that you own. This allows you to track changes, and deploy to cloud platforms
+easily.
+
+### Set up your API keys
+
+Make a copy of the `.env.example` file and update it with your own keys and
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
Make a copy of the `.env.example` file, rename it to `.env`, and update it with your own keys and
```

File: pulse/inngest-router/README.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
+# Prisma Pulse-Inngest Router
+
+This package enables event-driven support for your existing Inngest workflows,
```
Comment (jharrell) [inline pulse/inngest-router/README.md]: for `.md` files we generally don't line break like this. I'm assuming this is a Prettier thing? Would it be possible to edit the Prettier configuration to ignore `.md` files?

File: pulse/inngest-router/README.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+Make a copy of the `.env.example` file and update it with your own keys and
+database connection string:
+
+```bash
+cp .env.example .env
+```
+
+### Add your Prisma schema
+
+Add your existing Prisma schema file to the `/prisma` directory, overwriting
```
Comment (jharrell) [inline pulse/inngest-router/README.md]: This is a bit confusing to me. What if I don't have a Prisma Schema and I'm just trying things out?

File: pulse/inngest-router/README.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+docker run \
+  -e DATABASE_URL="" \
+  -e PULSE_API_KEY="" \
+  -e INNGEST_EVENT_KEY="" \
+  pulse-inngest-router
+```
+
+## Create Inngest functions
+
+When database changes occur, there will automatically be events submitted to
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
When database changes occur, there will be events automatically submitted to 
```

File: pulse/inngest-router/README.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+  { event: "db/create.notification" },
+  async ({ event, step }) => {
+    const pulseChange = event.data;
+
+    // Your logic here
+  };
+);
+```
+
+Note that the full [Pulse Change Event](https://www.prisma.io/docs/pulse/database-events) is available in `event.data`;
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
Note that the full [Pulse Change Event](https://www.prisma.io/docs/pulse/database-events) is available in `event.data`.
```

File: pulse/inngest-router/README.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+## Helpful documentation
+1. [Get started with Prisma ORM](https://www.prisma.io/docs/getting-started)
+1. [Get started with Prisma Pulse](https://www.prisma.io/docs/pulse)
+
+## Get help
+
+[Join our Discord community](https://discord.gg/KQyTW2H5ca) to get help with
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
[Join our Discord community](https://pris.ly/discord) to get help with
```

File: pulse/inngest-router/README.md (hunk 11 above)
Comment (sneub) [reply pulse/inngest-router/README.md]: No worries, fixed. It's an old habit of breaking lines at 80 characters (totally manually 😅)

File: pulse/inngest-router/README.md (hunk 12 above)
Comment (sneub) [reply pulse/inngest-router/README.md]: That's a very fair point. The original intent of this example was to be a deployable package for existing projects (just for context) which may explain a couple of the things.

But I think with some better explaining it should be OK

File: pulse/inngest-router/README.md [hunk 16]
```diff
@@ -0,0 +1,125 @@
+# Prisma Pulse-Inngest Router
+
+This package enables event-driven support for your existing Inngest workflows, using Prisma Pulse. Trigger workflows instantly whenever a database table changes, so you can create timely workflows that run instantly -- not only batch workflows.
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
This package enables event-driven support for your existing Inngest workflows, using Prisma Pulse. Trigger workflows whenever a database table changes, so you can create timely workflows that run instantly -- not only batch workflows.
```

File: pulse/inngest-router/README.md [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+```bash
+cp .env.example .env
+```
+
+### Add your Prisma schema
+
+Add your existing Prisma schema file to the `/prisma` directory, overwriting the existing example schema.
+
+If you don't already have a project using Inngest and Prisma Client, you may choose to try another [Prisma example](https://github.com/prisma/prisma-examples) instead.
```
Comment (jharrell) [inline pulse/inngest-router/README.md]: Up to you, but just to understand: 

Should we maybe rephrase this?
```suggestion
If you don't already have a project using Inngest and Prisma Client, you can choose another [Prisma example](https://github.com/prisma/prisma-examples) to start with instead.
```

File: pulse/inngest-router/README.md [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+
+Make a copy of the `.env.example` file, rename it to `.env`, and update it with your own keys and database connection string:
+
+```bash
+cp .env.example .env
+```
+
+### Add your Prisma schema
+
+Add your existing Prisma schema file to the `/prisma` directory, overwriting the existing example schema.
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
Add an existing Prisma schema file to the `/prisma` directory, overwriting the existing example schema.
```

File: pulse/inngest-router/README.md [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+
+If you don't already have a project using Inngest and Prisma Client, you may choose to try another [Prisma example](https://github.com/prisma/prisma-examples) instead.
+
+### Define the Prisma models to subscribe to
+
+Add all Prisma models that you wish to stream events from to the array. The service will ignore any models that don't exist.
```
Comment (jharrell) [inline pulse/inngest-router/README.md]:
```suggestion
Add all Prisma models you wish to stream events from into the array. The service will ignore any models that don't exist.
```

File: pulse/inngest-router/README.md [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+
+### Test locally before saving your changes
+
+Run the service locally:
+1. `npx prisma generate`
+1. `npm run build`
+1. `npm run start`
+
+**Hint:** you can trigger changes in your database by running
+`npx prisma studio` and modifying some data or adding a new row.
```
Comment (jharrell) [inline pulse/inngest-router/README.md]: Super unnecessary but you can use Github's admonition syntax here.
```suggestion
> [!TIP]
> you can trigger changes in your database by running `npx prisma studio` and modifying some data or adding a new row.
```

File: pulse/inngest-router/README.md [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+1. Create an account and project on [Railway](https://railway.app/new)
+1. Create a new service by clicking the **Create** button in the top right of
+your dashboard
+1. Choose to deploy from a **GitHub Repo**
+1. Authorize your GitHub account, and choose the GitHub repository you used in
+previous steps
+1. Add the respective environment variables:
+    1. DATABASE_URL
+    1. PULSE_API_KEY
+    1. INNGEST_EVENT_KEY
```
Comment (jharrell) [inline pulse/inngest-router/README.md]: Would change these to bullets instead of numbers (they don't have an inherent order)
```suggestion
   * DATABASE_URL
   * PULSE_API_KEY
   * INNGEST_EVENT_KEY
```

File: pulse/inngest-router/README.md (hunk 17 above)
Comment (jharrell) [reply pulse/inngest-router/README.md]: Actually, on second thought. Maybe we phrase this a different way since this repo has a sample `prisma.schema`
```suggestion
If you don't already have a project using Inngest and Prisma Client, you can use the sample Prisma Schema provided in this repo. Run `npx prisma db push` to push this schema to a test database of your choosing. [Check out the db push docs for more info](https://www.prisma.io/docs/orm/reference/prisma-cli-reference#db-push).

You can also try one of our other [Prisma examples](https://github.com/prisma/prisma-examples) as a starting point instead.
```

**Review by jharrell — DISMISSED** [review]

**Review by jharrell — APPROVED** [review]
